v23.1/transaction-retry-error-reference.md

```diff
@@ -1,54 +1,96 @@
+- If you're building an application with another driver or data access framework that is [supported by CockroachDB](third-party-database-tools.html), we recommend reusing the retry logic in our ["Simple CRUD" Example Apps](example-apps.html). For example, **Java** developers accessing the database with [JDBC](https://jdbc.postgresql.org) can reuse the example code implementing retry logic shown in [Build a Java app with CockroachDB](build-a-java-app-with-cockroachdb.html).
+- If you're building an application with a language and framework for which we do not provide example retry logic, you might need to write your own retry logic. For an example, see the [Client-side retry handling example](transaction-retry-error-example.html).
+- **Advanced users, such as library authors**: See [Advanced Client-Side Transaction Retries](advanced-client-side-transaction-retries.html).
+
+#### Client-side retry handling example
+
+For a conceptual example of application-defined retry logic, and testing that logic against your application's needs, see the [client-side retry handling example](transaction-retry-error-example.html).
+
+### Minimize transaction retry errors
+
+In addition to the steps described in [Client-side retry handling](#client-side-retry-handling), which detail how to configure your application to restart a failed transaction, there are also a number of changes you can make to your application logic to prevent client-side transaction retry errors from occurring in the first place:
+
+1. Limit the number of affected rows by following [performance-tuning best practices](apply-statement-performance-rules.html) (e.g., query performance tuning, index design and maintenance, etc.). Not only will transactions run faster and hold locks for a shorter duration, but the chances of [read invalidation](architecture/transaction-layer.html#read-refreshing) when the transaction’s [timestamp is pushed](architecture/transaction-layer.html#timestamp-cache) due to a conflicting write is decreased due to a smaller read set (i.e., a smaller number of rows read).
+
+1. Break down larger transactions into smaller ones (e.g., [bulk deletes](bulk-delete-data.html)) to have transactions hold locks for a shorter duration. This will also decrease the likelihood of [pushed timestamps](architecture/transaction-layer.html#timestamp-cache) and retry errors. For instance, as the size of writes (number of rows written) decreases, the chances of the (bulk delete) transaction’s timestamp getting bumped by concurrent reads decreases.
+
+1. Design your applications to reduce network round trips by [sending statements in transactions as a single batch](transactions.html#batched-statements) (e.g., using [common table expressions](common-table-expressions.html)).
```

Comment:
Can we elaborate why?

Batching allows CRDB to automaticall retry when previous reads are invalidated at a pushed timestamp.

https://www.cockroachlabs.com/docs/stable/transactions.html#automatic-retries

Reply:
Same reason as 4 below. The conditions for retry is less than 16kb and has not yet streamed results back. The later is why users should batch.

Reply:
Now #3 and #4 have nearly identical descriptions. To clarify then, can you elaborate on what you meant in your most recent comment, so I can add? 

> The conditions for retry is less than 16kb and has not yet streamed results back. The later is why users should batch.

Maybe (total guess):

> Batching allows CockroachDB to [automatically retry](transactions.html#automatic-retries) a transaction when [previous reads are invalidated](architecture/transaction-layer.html#read-refreshing) at a [pushed timestamp](architecture/transaction-layer.html#timestamp-cache), as long as the total transaction side (data written) is less than `16kb` and has not yet streamed any results back to the client.

Reply:
That's right! **Both** conditions (resultset size and less than 16KB) must be satisfied. 

For 3), If the user does not batch and they create a multi-statement transaction that does >1 roundtrip, CRDB will not automatically retry (even if the resultset is <16KB).

For 4), if the user sends their transaction as a single batch, but resultset hits 16KB where CRDB starts streaming results back, CRDB will not automatically retry.